File: LruCache.java
import java.util.*;
public class LruCache {
    HashMap<Integer, Node> map;
    Node head = new Node();
    Node tail = new Node();
    int capacity;

    public LruCache(int capacity) {
        map = new HashMap<>(capacity);
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        if (this.map.containsKey((key))) {
            Node node = this.map.get(key);
            remove(node);
            add(node);
            return node.val;
        }
        return -1;
    }
    
    public void put(int key, int value) {
        Node node = map.get(key);
        if (node != null) {
            remove(node);
            node.val = value;
            add(node);
        } else {
            if (map.size() == capacity) {
                map.remove(tail.prev.key);
                remove(tail.prev);
            }

            Node newNode = new Node();
            newNode.key = key;
            newNode.val = value;
            map.put(key, newNode);
            add(newNode);
        }
    }

    //To add node to the start of the list
    public void add(Node node) {
        Node headNext = head.next;
        head.next = node;
        node.prev = head;
        node.next = headNext;
        headNext.prev = node;
    }

    public void remove(Node node) {
        Node prevNode = node.prev;
        Node nextNode = node.next;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;
    }

    class Node {
        int key, val;
        Node prev, next;
    }
}

File: LruCache.js
var LRUCache = function(capacity) {
    this.capacity = capacity;
    this.map = new Map();
    this.head = new Node();
    this.tail = new Node();
    this.head.next = this.tail;
    this.tail.prev = this.head;
};


LRUCache.prototype.get = function(key) {
    if (!this.map.has(key)) {
        return -1;
    }
    const node = this.map.get(key)
    this.remove(node);
    this.insert(node);
    return node.val;
};


LRUCache.prototype.put = function(key, value) {
    if (this.map.has(key)) {
        var node = this.map.get(key);
        this.remove(node);
        node.val = value;
        this.insert(node);
    } else {
        if (this.map.size === this.capacity) {
            this.map.delete(this.tail.prev.key);
            this.remove(this.tail.prev);
        }
        const node = new Node(key, value);
        this.map.set(key, node);
        this.insert(node);
    }
};

LRUCache.prototype.insert = function(node) {
    const headNext = this.head.next;
    this.head.next = node;
    node.prev = this.head;
    node.next = headNext;
    headNext.prev = node;
}

LRUCache.prototype.remove = function(node) {
    const nodeNext = node.next;
    const nodePrev = node.prev;
    nodePrev.next = nodeNext;
    nodeNext.prev = nodePrev;
}

function Node(key, val, next, prev) {
    this.key = (key===undefined ? null : key);
    this.val = (val===undefined ? null : val);
    this.next = (next===undefined ? null : next);
    this.prev = (prev===undefined ? null : prev);
}

const cache = new LRUCache(2);
cache.put(1, 1);
cache.put(2, 2);
cache.put(3, 3);


